File: android/app/src/main/java/com/gestorfinances/app/domain/rules/RecurringAdvancer.kt
package com.gestorfinances.app.domain.rules

import java.time.LocalDate
import java.time.YearMonth

enum class RecurrenceFrequency {
    WEEKLY,
    FORTNIGHTLY,
    MONTHLY,
    YEARLY,
    CUSTOM,
}

enum class CustomRecurrenceUnit {
    DAYS,
    WEEKS,
    MONTHS,
    YEARS,
}

data class RecurrenceRule(
    val frequency: RecurrenceFrequency,
    val dayOfMonth: Int? = null,
    val intervalCount: Long? = null,
    val customUnit: CustomRecurrenceUnit? = null,
)

data class RecurrenceAdvance(
    val dueDates: List<LocalDate>,
    val newCursor: LocalDate,
)

object RecurringAdvancer {
    fun advance(
        rule: RecurrenceRule,
        cursor: LocalDate,
        today: LocalDate,
    ): RecurrenceAdvance {
        val dueDates = mutableListOf<LocalDate>()
        var next = cursor

        while (!next.isAfter(today)) {
            dueDates += next
            next = nextDate(rule, next)
        }

        return RecurrenceAdvance(dueDates = dueDates, newCursor = next)
    }

    private fun nextDate(rule: RecurrenceRule, current: LocalDate): LocalDate =
        when (rule.frequency) {
            RecurrenceFrequency.WEEKLY -> current.plusDays(7)
            RecurrenceFrequency.FORTNIGHTLY -> current.plusDays(14)
            RecurrenceFrequency.MONTHLY -> {
                val anchorDay = requireNotNull(rule.dayOfMonth) { "dayOfMonth is required" }
                clampDay(current.plusMonths(1), anchorDay)
            }
            RecurrenceFrequency.YEARLY -> {
                val anchorDay = requireNotNull(rule.dayOfMonth) { "dayOfMonth is required" }
                clampDay(current.plusYears(1), anchorDay)
            }
            RecurrenceFrequency.CUSTOM -> nextCustomDate(rule, current)
        }

    private fun nextCustomDate(rule: RecurrenceRule, current: LocalDate): LocalDate {
        val interval = requireNotNull(rule.intervalCount) { "intervalCount is required" }
        require(interval > 0) { "intervalCount must be positive" }

        return when (requireNotNull(rule.customUnit) { "customUnit is required" }) {
            CustomRecurrenceUnit.DAYS -> current.plusDays(interval)
            CustomRecurrenceUnit.WEEKS -> current.plusWeeks(interval)
            CustomRecurrenceUnit.MONTHS -> current.plusMonths(interval)
            CustomRecurrenceUnit.YEARS -> current.plusYears(interval)
        }
    }

    private fun clampDay(date: LocalDate, anchorDay: Int): LocalDate {
        require(anchorDay in 1..31) { "dayOfMonth must be 1..31" }
        val yearMonth = YearMonth.from(date)
        return yearMonth.atDay(anchorDay.coerceAtMost(yearMonth.lengthOfMonth()))
    }
}

File: windows/GestorFinances.Tests/Domain/Rules/RecurringAdvancer.cs
namespace GestorFinances.Domain.Rules;

public enum RecurrenceFrequency
{
    Weekly,
    Fortnightly,
    Monthly,
    Yearly,
    Custom
}

public enum CustomRecurrenceUnit
{
    Days,
    Weeks,
    Months,
    Years
}

public sealed record RecurrenceRule(
    RecurrenceFrequency Frequency,
    int? DayOfMonth = null,
    long? IntervalCount = null,
    CustomRecurrenceUnit? CustomUnit = null);

public sealed record RecurrenceAdvance(
    IReadOnlyList<DateOnly> DueDates,
    DateOnly NewCursor);

public static class RecurringAdvancer
{
    public static RecurrenceAdvance Advance(RecurrenceRule rule, DateOnly cursor, DateOnly today)
    {
        var dueDates = new List<DateOnly>();
        var next = cursor;

        while (next <= today)
        {
            dueDates.Add(next);
            next = NextDate(rule, next);
        }

        return new RecurrenceAdvance(dueDates, next);
    }

    private static DateOnly NextDate(RecurrenceRule rule, DateOnly current) =>
        rule.Frequency switch
        {
            RecurrenceFrequency.Weekly => current.AddDays(7),
            RecurrenceFrequency.Fortnightly => current.AddDays(14),
            RecurrenceFrequency.Monthly => ClampDay(current.AddMonths(1), RequiredDayOfMonth(rule)),
            RecurrenceFrequency.Yearly => ClampDay(current.AddYears(1), RequiredDayOfMonth(rule)),
            RecurrenceFrequency.Custom => NextCustomDate(rule, current),
            _ => throw new ArgumentOutOfRangeException(nameof(rule), "Unknown recurrence frequency")
        };

    private static DateOnly NextCustomDate(RecurrenceRule rule, DateOnly current)
    {
        var interval = rule.IntervalCount ?? throw new ArgumentException("intervalCount is required", nameof(rule));
        if (interval <= 0)
        {
            throw new ArgumentException("intervalCount must be positive", nameof(rule));
        }

        return (rule.CustomUnit ?? throw new ArgumentException("customUnit is required", nameof(rule))) switch
        {
            CustomRecurrenceUnit.Days => current.AddDays((int)interval),
            CustomRecurrenceUnit.Weeks => current.AddDays((int)(interval * 7)),
            CustomRecurrenceUnit.Months => current.AddMonths((int)interval),
            CustomRecurrenceUnit.Years => current.AddYears((int)interval),
            _ => throw new ArgumentOutOfRangeException(nameof(rule), "Unknown custom recurrence unit")
        };
    }

    private static int RequiredDayOfMonth(RecurrenceRule rule) =>
        rule.DayOfMonth ?? throw new ArgumentException("dayOfMonth is required", nameof(rule));

    private static DateOnly ClampDay(DateOnly date, int anchorDay)
    {
        if (anchorDay is < 1 or > 31)
        {
            throw new ArgumentException("dayOfMonth must be 1..31", nameof(anchorDay));
        }

        var day = Math.Min(anchorDay, DateTime.DaysInMonth(date.Year, date.Month));
        return new DateOnly(date.Year, date.Month, day);
    }
}
